runtime/ints.c

```diff
@@ -604,7 +604,7 @@ CAMLprim value caml_int64_of_string(value s)
       if (res >  (uint64_t)1 << 63) caml_failwith(INT64_ERRMSG);
     }
   }
-  if (sign < 0) res = - res;
+  if (sign < 0) res = ~res + 1;
```

Comment:
Not really fond of this. Wouldn't `res = (uint64_t)-res` be good enough to silence those warnings?

Reply:
`res` is already `uint64_t`. Not sure how `(uint64_t)-(int64_t)res` really behaves.
`res = UINT64_MAX - res + 1` isn't an improvement.
What if we went with a comment? `res = ~res + 1 /* res = -res */`

Reply:
If `res` is already `int64_t`, then `(uint64_t)-res` will work nicely, since its value has already been bounds checked above so we know `0 <= res <= INT64_MAX`.

Reply:
I think you've misread, `res` is unsigned (`uint64_t`). The compiler warning is probably overzealous, but the compiler doesn't like when an unary minus is applied to an unsigned integer.

Reply:
Ah, so two casts. Ugly but still more human-friendly than bitmasking.

Reply:
Turns out a single cast is sufficient.